src/3d/qgs3dmapcanvas.cpp

```diff
@@ -348,3 +351,60 @@ QVector<QgsPointXY> Qgs3DMapCanvas::viewFrustum2DExtent()
 {
   return mScene ? mScene->viewFrustum2DExtent() : QVector<QgsPointXY>();
 }
+
+void Qgs3DMapCanvas::highlightFeature( const QgsFeature &feature, QgsMapLayer *layer )
+{
+  // we only support point clouds for now
+  if ( layer->type() != Qgis::LayerType::PointCloud )
+    return;
+
+  const QgsGeometry geom = feature.geometry();
+  QgsPoint pt( geom.vertexAt( 0 ) );
+  pt.setZ( pt.z() / mMapSettings->terrainVerticalScale() );
```

Comment:
hmm this is a bit tricky - point's Z value may not depend on terrain's vertical scale (and point clouds are never clamped to terrain).  I would skip this Z update, and then for vector layers it can be adjusted based on how the Z value gets calculated...

Reply:
This is weird, I copied this logic from the 3d measure tool and they both work as expected.

It seems there's a bug in `Qgs3DUtils::clampAltitude` and `Qgs3DUtils::clampAltitudes` functions:
`AltitudeClamping::Absolute` is treated the same as `AltitudeClamping::Relative` so terrainZ and vertical scale are always applied.

I'll skip the Z update here but it will not render correctly when terrain vertical scale is set, until we fix the utils functions.